pkg/vmssrollout/upgrade.go

```diff
@@ -0,0 +1,342 @@
+		if u.Drain {
+			log.Printf("Drain VM %q in VMSS %q", *vmToUpgrade.InstanceID, u.Name)
+			nodeName := *vmToUpgrade.VirtualMachineScaleSetVMProperties.OsProfile.ComputerName
+			if err := u.drain(nodeName); err != nil {
+				return err
+			}
+		}
+
+		// Re-image VMs only on in-place upgrades. Otherwise, we have already
+		// scaled up and need to scale down old VMs.
+		if u.InPlace && u.reimageOnly() {
+			log.Printf("Reimage VM %q in VMSS %q", *vmToUpgrade.InstanceID, u.Name)
+			update := compute.VirtualMachineScaleSetVMInstanceIDs{
+				InstanceIds: &[]string{*vmToUpgrade.InstanceID},
+			}
+			res, err := u.ssc.Reimage(context.Background(), u.ResourceGroup, u.Name, &update)
+			if err != nil {
+				return fmt.Errorf("cannot update VM %q: %v", *vmToUpgrade.InstanceID, err)
+			}
+			// TODO: Uncordon VM if we drained it.
+			if err := res.WaitForCompletion(context.Background(), u.vmc.BaseClient.Client); err != nil {
+				return fmt.Errorf("cannot wait for VM %q reimage: %v", *vmToUpgrade.InstanceID, err)
+			}
+			if _, err := res.Result(u.ssc); err != nil {
+				return fmt.Errorf("error during VM %q reimage: %v", *vmToUpgrade.InstanceID, err)
+			}
+			// Wait for VM readiness; this may vary between VMSSs, eg. a master VM
+			// is ready when the API server pod running on it is ready whereas a compute
+			// VM is ready when it has joined the cluster as a Ready k8s node.
+			nodeName := *vmToUpgrade.VirtualMachineScaleSetVMProperties.OsProfile.ComputerName
+			if err := u.waitForReady(nodeName); err != nil {
+				return err
+			}
+			continue
+		}
+
+		if u.InPlace {
+			log.Printf("In-place upgrade VM %q in VMSS %q", *vmToUpgrade.InstanceID, u.Name)
+			update := compute.VirtualMachineScaleSetVMInstanceRequiredIDs{
+				InstanceIds: &[]string{*vmToUpgrade.InstanceID},
+			}
+			res, err := u.ssc.UpdateInstances(context.Background(), u.ResourceGroup, u.Name, update)
+			if err != nil {
+				return fmt.Errorf("cannot update VM %q: %v", *vmToUpgrade.InstanceID, err)
+			}
+			// TODO: Uncordon VM if we drained it.
+			if err := res.WaitForCompletion(context.Background(), u.vmc.BaseClient.Client); err != nil {
+				return fmt.Errorf("cannot wait for VM %q update: %v", *vmToUpgrade.InstanceID, err)
+			}
+			if _, err := res.Result(u.ssc); err != nil {
+				return fmt.Errorf("error during VM %q update: %v", *vmToUpgrade.InstanceID, err)
+			}
+			// Wait for VM readiness; this may vary between VMSSs, eg. a master VM
+			// is ready when the API server pod running on it is ready whereas a compute
+			// VM is ready when it has joined the cluster as a Ready k8s node.
+			nodeName := *vmToUpgrade.VirtualMachineScaleSetVMProperties.OsProfile.ComputerName
+			if err := u.waitForReady(nodeName); err != nil {
+				return err
+			}
+			continue
+		}
+
+		log.Printf("Deleting VM %q in VMSS %q", *vmToUpgrade.InstanceID, u.Name)
+		res, err := u.vmc.Delete(context.Background(), u.ResourceGroup, u.Name, *vmToUpgrade.InstanceID)
```

Comment:
If we did InPace upgrade, why we should want to delete node?

Reply:
we don't delete on in-place

Reply:
And how azure behaves if you scale up to 4, and delete 1? will it initiates 1 creation again or its just leaves 3?

Reply:
VMSS is resized to 3.